common/src/database/redis.rs

```diff
@@ -0,0 +1,794 @@
+        }
+
+        let key = format!("{}:{}:{}", RedisDomain::Subscription, subscription.uuid().to_string().to_uppercase(), subscription.name());
+        let value = serde_json::to_string(subscription).context("Failed to serialize subscription data")?;
+        let _ : String = conn.set(key, value).await.context("Failed to store subscription data")?;
+        Ok(())
+    }
+
+    async fn delete_subscription(&self, uuid: &str) -> Result<()> {
+        let mut conn = self.pool.get().await.context("Failed to get Redis connection")?;
+        let first_pass_key = format!("{}:{}:{}", RedisDomain::Subscription, uuid.to_uppercase(), RedisDomain::Any);
+        let second_pass_key = format!("{}:{}:{}", RedisDomain::Subscription, RedisDomain::Any, uuid);
+
+        let keys = list_keys_with_fallback(&mut conn, &first_pass_key, &second_pass_key).await?;
+
+        self.delete_bookmarks(None, Some(uuid)).await.context("Failed to delete subscription releated bookmark data")?;
+        if !keys.is_empty() {
+            let _: () = conn.del(keys).await.context("Failed to delete subscription data")?;
+        }
+        Ok(())
+    }
+
+    /// Fails if `setup_schema` hasn't previously been called or if the query otherwise fails.
+    async fn current_version(&self) -> Result<Option<Version>> {
+        let mut conn = self.pool.get().await.context("Failed to get Redis connection")?;
+        let key = MIGRATION_TABLE_NAME;
+        let versions:Vec<String> = conn.zrange(key, -1, -1).await.context("There is no version info stored in DB.")?;
+        let last_version = versions.last().and_then(|v| v.parse::<i64>().ok());
+        Ok(last_version)
+    }
+
+    /// Fails if `setup_schema` hasn't previously been called or if the query otherwise fails.
+    async fn migrated_versions(&self) -> Result<BTreeSet<Version>> {
+        let mut conn = self.pool.get().await.context("Failed to get Redis connection")?;
+        let key = MIGRATION_TABLE_NAME;
+        let versions:Vec<String> = conn.zrange(key, 0, -1).await.context("There is no version info stored in DB.")?;
+        let result : BTreeSet<i64> = versions.into_iter().map(|v| v.parse::<i64>().context(format!("Failed to parse version: {}", v))).collect::<Result<_>>()?;
+        Ok(result)
+    }
+
+    /// Fails if `setup_schema` hasn't previously been called or if the migration otherwise fails.
+    async fn apply_migration(&self, version: Version) -> Result<()> {
+        let migration = self
+            .migrations
+            .get(&version)
+            .ok_or_else(|| anyhow!("Could not retrieve migration with version {}", version))?
+            .clone();
+        let mut conn = self.pool.get().await.context("Failed to get Redis connection")?;
+        migration.up(&mut conn).await?;
+        let key = MIGRATION_TABLE_NAME;
+        let version = migration.version();
+        let added_count: i64 = conn.zadd(key, version, version).await.context(format!("Unable to add version: {}", version))?;
```

Comment:
It would be better to use `with_context` instead of `context` since it does not just contain a static str.